### PR TITLE
Users guide: zippered iteration

### PR DESCRIPTION
**Pull request by bradcray**
This is a proposed draft of a users guide section on zippered iteration plus a supporting test

### CODE REVIEW

File: doc/sphinx/source/users-guide/base/zip.rst [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  :language: chapel
+  :lines: 12-13
+
+The output of this loop is:
+
+.. literalinclude:: examples/users-guide/base/zip.good
+  :language: text
+  :lines: 5-9
+
+In fact, the original case can be thought of as de-tupling the
```
Comment (psahabu) [inline doc/sphinx/source/users-guide/base/zip.rst]: Could also say "de-structuring" or "pattern matching", but maybe better left for "Chapel for Functional Programmers".

File: doc/sphinx/source/users-guide/base/zip.rst [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+their ordinal values without needing to know the array's size *a
+priori*:
+
+.. literalinclude:: examples/users-guide/base/zip.chpl
+  :language: chapel
+  :lines: 37-38
+
+At present, there is unfortunately no way for users to define their
+own iterator functions in a way that permits them to be flexible with
+respect to size like unbounded ranges are.  However, this is planned
```
Comment (psahabu) [inline doc/sphinx/source/users-guide/base/zip.rst]: Awkward wording here, suggested edit.

"However, we plan to support such idioms in the future, like consuming n values from a random sequence or a data stream in a zippered context."

File: doc/sphinx/source/users-guide/base/zip.rst [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+for the future to support idioms like consuming *n* values from a
+random sequence or a data stream in a zippered context.  In the
+meantime, a typical workaround is to write such iterators to take the
+number of items to generate as an argument.
+
+Related: Today, unbounded range iterands can occur anywhere in a
```
Comment (psahabu) [inline doc/sphinx/source/users-guide/base/zip.rst]: Maybe italicize "Related:" to draw more attention.

**Comment by psahabu** [comment]
Otherwise looks good.

**Comment by bradcray** [comment]
@psahabu : Thanks for the review.  I made some changes based on your recommendations and a few others based on having taken a break from it.  Let me know if you want to review further or if I'm OK to merge.

**Comment by psahabu** [comment]
Looks good to me.

**Pull request merged**